File: Design-Patterns-Using-Cpp/Composite/CompositePattern.cpp
#include <iostream>
#include <unordered_set>

// Component
class Component {
public:
    virtual void operation() = 0;
    virtual ~Component() = default;
};

// Composite
class Composite : public Component {
private:
    std::unordered_set<Component*> children;

public:
    void operation() override {
        std::cout << "Composite Operation" << std::endl;
        for (Component* child : children) {
            child->operation();
        }
    }

    void add(Component* component) {
        children.insert(component);
    }

    void remove(Component* component) {
        children.erase(component);
    }
};

// Leaf
class Leaf : public Component {
public:
    void operation() override {
        std::cout << "Leaf Operation" << std::endl;
    }
};

// Client code
int main() {
    Composite composite;
    composite.add(new Leaf());

    Composite composite2;
    composite2.add(new Leaf());

    composite.add(&composite2);
    composite.operation();

    return 0;
}

/*
Composite Operation
Composite Operation
Leaf Operation
Leaf Operation
*/

File: Design-Patterns-Using-Swift/Composite/CompositePattern.swift
// Component
class Component: Hashable {
    static func == (lhs: Component, rhs: Component) -> Bool {
        return ObjectIdentifier(lhs) == ObjectIdentifier(rhs)
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(ObjectIdentifier(self))
    }
    
    func operation() {
        fatalError("Operation must be overridden by subclasses")
    }
}

// Composite
class Composite: Component {
    private var children = Set<Component>()

    override func operation() {
        print("Composite Operation")
        for child in children {
            child.operation()
        }
    }

    func add(_ component: Component) {
        children.insert(component)
    }

    func remove(_ component: Component) {
        children.remove(component)
    }
}

// Leaf
class Leaf: Component {
    override func operation() {
        print("Leaf Operation")
    }
}

// Client code
let leaf1 = Leaf()
let leaf2 = Leaf()

let composite = Composite()
composite.add(leaf1)

let composite2 = Composite()
composite2.add(leaf2)

composite.add(composite2)

composite.operation()

/*
Composite Operation
Leaf Operation
Composite Operation
Leaf Operation
*/